src/compiler/program.ts

```diff
@@ -773,10 +784,30 @@ namespace ts {
             const modifiedSourceFiles: { oldFile: SourceFile, newFile: SourceFile }[] = [];
             oldProgram.structureIsReused = StructureIsReused.Completely;
 
-            for (const oldSourceFile of oldProgram.getSourceFiles()) {
+            const oldSourceFiles = oldProgram.getSourceFiles();
+            const enum SeenPackageName { Exists, Modified }
+            const seenPackageNames = createMap<SeenPackageName>();
+
+            for (const oldSourceFile of oldSourceFiles) {
                 const newSourceFile = host.getSourceFileByPath
                     ? host.getSourceFileByPath(oldSourceFile.fileName, oldSourceFile.path, options.target)
                     : host.getSourceFile(oldSourceFile.fileName, options.target);
+                Debug.assert(!(newSourceFile && newSourceFile.redirect), "Host should not return a redirect source file from `getSourceFile`");
+
+                if (oldSourceFile.redirect) {
+                    // We got `newSourceFile` by path, so it is actually for the underlying file.
+                    // This lets us know if the underlying file has changed. If it has we should break the redirect.
+                    if (newSourceFile !== oldSourceFile.redirect.underlying) {
+                        // Underlying file has changed. Might not redirect anymore. Must rebuild program.
+                        return oldProgram.structureIsReused = StructureIsReused.Not;

```

Comment:
Per our offline discussion, I agree that this is an appropriate way to handle changes.